### ISSUE TITLE
Embedding loading slow

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#13472]
**Describe the bug**
Loading an embedding is a lot slower on 0.35.0+, and when there's multiple embeds on a page, then it blocks loading for many seconds without showing any indication (spinner etc).
I'm not sure if this is caused by pure JS blockers, or if this is something in the connection handling (Jetty etc). I have tried to debug, but got lost. At least I have found when the problem started.

**To Reproduce**
1. Create 8 questions, with embedding enabled. Something simple like `SELECT 1` as "Q1", `SELECT 2` as "Q2" etc
2. Example embed loader (rename to `.html`) and update all the embed URLs
<details><summary>Show HTML code</summary>

```
﻿<!DOCTYPE html>
<html>
<head>
    <title>Metabase Embedding Test</title>
</head>
<body>

<script type="text/javascript">
var startTime = Date.now();
var interval = setInterval(function() {
    var elapsedTime = Date.now() - startTime;
    document.getElementById("timer").innerHTML = (elapsedTime / 1000).toFixed(1);
}, 100);
window.onload = function() { clearInterval(interval); }
</script>
<h1 id="timer"></h1>

<!-- ->

<script type="text/javascript">
  window.onload = function(){
    for (let step = 0; step < 8; step++) {
        var f = document.createElement('iframe');
        f.src = "http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6Mjk2fSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ0ODMxLCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.V20fOWJ1Se9qeNd2lpORHuEMLREkJOE2Nx7fWMg10BY#bordered=true&titled=true";
        f.width = "400"; 
        f.height = "300";
        f.frameborder = "0";
        f.allowtransparency = "true";
        document.body.appendChild(f);
    }
  }
</script>

<!-- -->

<iframe src="http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6Mjk2fSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ0ODMxLCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.V20fOWJ1Se9qeNd2lpORHuEMLREkJOE2Nx7fWMg10BY#bordered=true&titled=true"
        style="border:5px solid red;" onload="this.style.border='5px solid green';"
        width="400"
        height="300"
        allowtransparency></iframe>

<!-- -->

<iframe src="http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6Mjk3fSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ0ODg4LCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.XR744oNUn6sEiLmqGKKuNkLPlLF6OEgrxmAnPy1U6ZM#bordered=true&titled=true"
        style="border:5px solid red;" onload="this.style.border='5px solid green';"
        width="400"
        height="300"
        allowtransparency></iframe>

<iframe src="http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6Mjk4fSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ0OTQzLCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.TPuKIsPht1cZ0Z5c_lD_KbYnSO573l9_kr3pcOr32cI#bordered=true&titled=true"
        style="border:5px solid red;" onload="this.style.border='5px solid green';"
        width="400"
        height="300"
        allowtransparency></iframe>

<iframe src="http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6Mjk5fSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ0OTczLCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.SxQA96xEHOXIuIaNCjoYyaOJva80w95PkVmiXRrhL2I#bordered=true&titled=true"
        style="border:5px solid red;" onload="this.style.border='5px solid green';"
        width="400"
        height="300"
        allowtransparency></iframe>

<iframe src="http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6MzAwfSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ1MDAzLCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.zcDkxlL_hPvSkI5uAzxovxKqa4QY2QrrscfR7eM64m0#bordered=true&titled=true"
        style="border:5px solid red;" onload="this.style.border='5px solid green';"
        width="400"
        height="300"
        allowtransparency></iframe>

<iframe src="http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6MzAxfSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ1MDM2LCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.zawd6wwzv0SEzUHfrOtI1SjpDHdatLKUlZEOUBPV5Cs#bordered=true&titled=true"
        style="border:5px solid red;" onload="this.style.border='5px solid green';"
        width="400"
        height="300"
        allowtransparency></iframe>

<iframe src="http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6MzAyfSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ1MDY0LCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.QD6onB-bYtrEPkIqpPVnnCQWR9JgqpMMJH1zFX_nYDQ#bordered=true&titled=true"
        style="border:5px solid red;" onload="this.style.border='5px solid green';"
        width="400"
        height="300"
        allowtransparency></iframe>

<iframe src="http://localhost:3001/embed/question/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyZXNvdXJjZSI6eyJxdWVzdGlvbiI6MzAzfSwicGFyYW1zIjp7fSwiaWF0IjoxNjAxNjQ1MTMwLCJfZW1iZWRkaW5nX3BhcmFtcyI6e319.SdZYxiDyHjXk0gRbo9Q6q_79gAFaddEYUoTdRB3WZD0#bordered=true&titled=true"
        style="border:5px solid red;" onload="this.style.border='5px solid green';"
        width="400"
        height="300"
        allowtransparency></iframe>

<!-- -->

</body>
</html>
```

</details>

3. Run the test on 0.35.0 (or any version above, I have tested up to 0.37.0-rc2)
![Screencast-2020-10-14_17 46 24](https://user-images.githubusercontent.com/1447303/96013533-aa710d00-0e45-11eb-9490-7417011c9451.gif)
   - 1 embed will load in ~2.5 seconds, with visual spinner maybe showing .1 seconds before load.
   - 8 embeds will load in ~19 seconds, with no indication of what is happening until about 1 second before the end.
4. Run the test on 0.34.3
![Screencast-2020-10-14_17 44 03](https://user-images.githubusercontent.com/1447303/96013518-a513c280-0e45-11eb-8bd4-8fcd71ce3e07.gif)
   - 1 embed will load in ~0.9 seconds, with visual spinner maybe showing .1 seconds before load.
   - 8 embeds will load in ~5 seconds, with visual spinner showing on all embeds around 4 second mark.

Another way to "reproduce":
Simply go to the embed preview in 0.34.3, which will load the preview much faster than any newer versions.

**Expected behavior**
At least non-blocking behavior, so a spinner is shown. Preferably sub-1-second load time.

**Comment by salsakran** [comment]
Since this is both a regression, and a brutal slowdown (19s for 8 cards is not ok) I'm bumping this to a P1

**Comment by ariya** [comment]
Thank you @flamber for such a detailed analysis!

I followed up your track and bisected all the ~200 commits between v0.34 and v0.35. Finally, I managed to narrow it down to single commit, 09001d6c5, that started to cause the embedding speed regression.

Looking at the front-end code difference between v0.34 and commit 09001d6c5 (and effectively, v0.35 or later), the JavaScript grows by approximately 280KB. Here is the breakdown:

In v0.34:

```
 3576195 resources/frontend_client/app/dist/vendor.bundle.js
    2487 resources/frontend_client/app/dist/styles.bundle.js
 1345125 resources/frontend_client/app/dist/app-embed.bundle.js
   13084 resources/frontend_client/app/dist/0.bundle.js
 4936891 TOTAL (4.71 MB)
```

In v0.35 (right after commit 09001d6c5):

```
 3763334 resources/frontend_client/app/dist/vendor.bundle.js
    2510 resources/frontend_client/app/dist/styles.bundle.js
 1460501 resources/frontend_client/app/dist/app-embed.bundle.js
 5226345 TOTAL (4.98 KB)
```

The above size difference causes the trigger to fetch the JSON (via XHR) to happen at different times as well. Notionly, it is a bit later for v0.35.

To demonstrate this, I captured the network waterfall diagram (see the attached HARs, can be displayed with e.g.[Online HAR Viewer](http://www.softwareishard.com/har/viewer/)) for both v0.34 and v0.35. To further stress the front-end code, this was carried out under the "Fast 3G" network throttling simulation in Chrome Dev Tools.

[v34.har.txt](https://github.com/metabase/metabase/files/5698489/v34.har.txt)
[v35.har.txt](https://github.com/metabase/metabase/files/5698490/v35.har.txt)

With v0.34, the fetch is initiated after 8.69 seconds.

![image](https://user-images.githubusercontent.com/7288/102270394-c854ff80-3ed2-11eb-9292-09d857ae1634.png)

However, with v0.35, it happens almost a full second later, at 9.60 seconds.

![image](https://user-images.githubusercontent.com/7288/102270420-cf7c0d80-3ed2-11eb-94f7-e6e0adc56211.png)

This delay contributes to the slow down of iframe rendering for this embedding use case. Note that this is just testing a single (embedded) iframe. With a few more, there will be a cascading effect, adding further to the overall slow experience.

My next attemp is to figure out why this extra 280 KB of JavaScript code adds so much delay (~1 s) to the initial data fetch and if there is a way to workaround it.

**Comment by ariya** [comment]
For completeness, here is the code inflation due to the said commit, restricted to `frontend/src`:

<table>
<thead><tr><td>File</td><td>Before</td><td>After</td><td>Delta</td></tr></thead>
<tr><td>metabase-lib/lib/queries/StructuredQuery.js</td><td align=right>42916</td><td align=right>43211</td><td align=right>295</td></tr>
<tr><td>metabase-lib/lib/queries/structured/Aggregation.js</td><td align=right>6821</td><td align=right>6813</td><td align=right>-8</td></tr>
<tr><td>metabase-lib/lib/queries/structured/Filter.js</td><td align=right>8295</td><td align=right>8434</td><td align=right>139</td></tr>
<tr><td>metabase-lib/lib/queries/structured/MBQLClause.js</td><td align=right>2472</td><td align=right>2529</td><td align=right>57</td></tr>
<tr><td>metabase/lib/expressions/compile.js</td><td align=right>0</td><td align=right>5414</td><td align=right>5414</td></tr>
<tr><td>metabase/lib/expressions/config.js</td><td align=right>560</td><td align=right>8181</td><td align=right>7621</td></tr>
<tr><td>metabase/lib/expressions/format.js</td><td align=right>0</td><td align=right>3927</td><td align=right>3927</td></tr>
<tr><td>metabase/lib/expressions/formatter.js</td><td align=right>2155</td><td align=right>0</td><td align=right>-2155</td></tr>
<tr><td>metabase/lib/expressions/index.js</td><td align=right>2148</td><td align=right>4787</td><td align=right>2639</td></tr>
<tr><td>metabase/lib/expressions/lexer.js</td><td align=right>0</td><td align=right>7800</td><td align=right>7800</td></tr>
<tr><td>metabase/lib/expressions/parser.js</td><td align=right>16306</td><td align=right>13146</td><td align=right>-3160</td></tr>
<tr><td>metabase/lib/expressions/process.js</td><td align=right>0</td><td align=right>1291</td><td align=right>1291</td></tr>
<tr><td>metabase/lib/expressions/suggest.js</td><td align=right>0</td><td align=right>15513</td><td align=right>15513</td></tr>
<tr><td>metabase/lib/expressions/syntax.js</td><td align=right>0</td><td align=right>10224</td><td align=right>10224</td></tr>
<tr><td>metabase/lib/expressions/tokens.js</td><td align=right>2308</td><td align=right>0</td><td align=right>-2308</td></tr>
<tr><td>metabase/lib/query.js</td><td align=right>13634</td><td align=right>13631</td><td align=right>-3</td></tr>
<tr><td>metabase/lib/query/field_ref.js</td><td align=right>5310</td><td align=right>5539</td><td align=right>229</td></tr>
<tr><td>metabase/lib/query/filter.js</td><td align=right>3080</td><td align=right>3117</td><td align=right>37</td></tr>
<tr><td>metabase/lib/query_aggregation.js</td><td align=right>3190</td><td align=right>3078</td><td align=right>-112</td></tr>
<tr><td>metabase/lib/schema_metadata.js</td><td align=right>18048</td><td align=right>18152</td><td align=right>104</td></tr>
<tr><td>metabase/modes/components/drill/UnderlyingRecordsDrill.jsx</td><td align=right>1333</td><td align=right>2234</td><td align=right>901</td></tr>
<tr><td>metabase/query_builder/components/AggregationName.jsx</td><td align=right>3870</td><td align=right>3921</td><td align=right>51</td></tr>
<tr><td>metabase/query_builder/components/AggregationPopover.jsx</td><td align=right>13874</td><td align=right>12115</td><td align=right>-1759</td></tr>
<tr><td>metabase/query_builder/components/DimensionList.jsx</td><td align=right>7687</td><td align=right>7707</td><td align=right>20</td></tr>
<tr><td>metabase/query_builder/components/ExpressionPopover.jsx</td><td align=right>0</td><td align=right>2601</td><td align=right>2601</td></tr>
<tr><td>metabase/query_builder/components/ExtendedOptions.jsx</td><td align=right>6669</td><td align=right>6611</td><td align=right>-58</td></tr>
<tr><td>metabase/query_builder/components/FieldList.jsx</td><td align=right>2919</td><td align=right>2934</td><td align=right>15</td></tr>
<tr><td>metabase/query_builder/components/expressions/ExpressionEditorTextfield.jsx</td><td align=right>11473</td><td align=right>11241</td><td align=right>-232</td></tr>
<tr><td>metabase/query_builder/components/expressions/ExpressionWidget.jsx</td><td align=right>4046</td><td align=right>4254</td><td align=right>208</td></tr>
<tr><td>metabase/query_builder/components/expressions/Expressions.jsx</td><td align=right>2079</td><td align=right>1959</td><td align=right>-120</td></tr>
<tr><td>metabase/query_builder/components/expressions/TokenizedExpression.jsx</td><td align=right>3366</td><td align=right>703</td><td align=right>-2663</td></tr>
<tr><td>metabase/query_builder/components/expressions/TokenizedInput.jsx</td><td align=right>4926</td><td align=right>5157</td><td align=right>231</td></tr>
<tr><td>metabase/query_builder/components/filters/FilterPopover.jsx</td><td align=right>5571</td><td align=right>7098</td><td align=right>1527</td></tr>
<tr><td>metabase/query_builder/components/filters/FilterPopoverHeader.jsx</td><td align=right>1608</td><td align=right>1565</td><td align=right>-43</td></tr>
<tr><td>metabase/query_builder/components/filters/FilterPopoverPicker.jsx</td><td align=right>1312</td><td align=right>1786</td><td align=right>474</td></tr>
<tr><td>TOTAL</td><td align=right>197976</td><td align=right>246673</td><td align=right>48697</td></tr>
</table>

The bulk of it is concentrated inside `metabase/lib/expressions`, therefore if we zoom in there:

<table>
<thead><tr><td>File</td><td>Before</td><td>After</td><td>Delta</td></tr></thead>
<tr><td>metabase/lib/expressions/compile.js</td><td align=right>0</td><td align=right>5414</td><td align=right>5414</td></tr>
<tr><td>metabase/lib/expressions/config.js</td><td align=right>560</td><td align=right>8181</td><td align=right>7621</td></tr>
<tr><td>metabase/lib/expressions/format.js</td><td align=right>0</td><td align=right>3927</td><td align=right>3927</td></tr>
<tr><td>metabase/lib/expressions/formatter.js</td><td align=right>2155</td><td align=right>0</td><td align=right>-2155</td></tr>
<tr><td>metabase/lib/expressions/index.js</td><td align=right>2148</td><td align=right>4787</td><td align=right>2639</td></tr>
<tr><td>metabase/lib/expressions/lexer.js</td><td align=right>0</td><td align=right>7800</td><td align=right>7800</td></tr>
<tr><td>metabase/lib/expressions/parser.js</td><td align=right>16306</td><td align=right>13146</td><td align=right>-3160</td></tr>
<tr><td>metabase/lib/expressions/process.js</td><td align=right>0</td><td align=right>1291</td><td align=right>1291</td></tr>
<tr><td>metabase/lib/expressions/suggest.js</td><td align=right>0</td><td align=right>15513</td><td align=right>15513</td></tr>
<tr><td>metabase/lib/expressions/syntax.js</td><td align=right>0</td><td align=right>10224</td><td align=right>10224</td></tr>
<tr><td>metabase/lib/expressions/tokens.js</td><td align=right>2308</td><td align=right>0</td><td align=right>-2308</td></tr>
<tr><td>metabase/lib/query.js</td><td align=right>13634</td><td align=right>13631</td><td align=right>-3</td></tr>
<tr><td>metabase/lib/query/field_ref.js</td><td align=right>5310</td><td align=right>5539</td><td align=right>229</td></tr>
<tr><td>metabase/lib/query/filter.js</td><td align=right>3080</td><td align=right>3117</td><td align=right>37</td></tr>
<tr><td>metabase/lib/query_aggregation.js</td><td align=right>3190</td><td align=right>3078</td><td align=right>-112</td></tr>
<tr><td>metabase/lib/schema_metadata.js</td><td align=right>18048</td><td align=right>18152</td><td align=right>104</td></tr>
<tr><td>TOTAL</td><td align=right>66739</td><td align=right>113800</td><td align=right>47061</td></tr>
</table>

My next step is find out why this extra 45 KB code adds so much to the initial delay of the (embedding) app.

**Comment by ariya** [comment]
Another quick update.

I can now see an obvious place where the said commit introduces a new significant JavaScript code execution.

_Before_ the commit, an illustrative example of the call tree from the code execution flame chart, before DOMContentLoaded event (which happens at 550 ms):

![image](https://user-images.githubusercontent.com/7288/102560345-df3c5300-4086-11eb-87b0-d757b9a11d78.png)

Chrome DevTools estimated the total blocking time to be ~167 ms.

In the meantime, _after_ the commit was introduced (look at that deep, nested call tree!):

![image](https://user-images.githubusercontent.com/7288/102560232-9c7a7b00-4086-11eb-9ad4-88b6ca10721d.png)

which contributed to a whopping (estimated) 730 ms blocking time. This also explains why DOMContentLoaded event fires after 1.12 seconds.

Next step: navigate the flame chart, understand the breakdown of the slow down even more.

**Comment by ariya** [comment]
I can now confirm that the slow-down is caused by Chevrotain's `performSelfAnalysis`, which needs to be called in the construction of the expression parsers (there are 3 instances of different parsers). In fact, this has nothing to do with embedding at all! The problem already manifests itself in the regular Metabase app, e.g. by going to the Sign In screen (however, the slow down is much much more perceptible when doing embedding, since there could be multiple iframes at once).

To reproduce:

1. Launch Metabase, go to the Sign In screen
2. Open Chrome DevTools, _Performance_ panel
3. Click _Record_, reload the page and once the page is completed, click _Stop_
4. Drill-down the flame chart to locate the execution tree

As shown in the following screenshot, `performSelfAnalysis` is rather expensive (220 ms, and I'm on a fast 3.6 GHz desktop rig!) and it must be carried out for all the 3 different parsers. Because these steps happen in the beginning (application initialization), it delays the DOMContentLoaded up to 700 ms (or more, on a slower machine).

![image](https://user-images.githubusercontent.com/7288/102634598-e3a85080-4106-11eb-80ed-ed4351920ae7.png)

Next step: find a way to avoid this up-front expensive initialization.

**Comment by ariya** [comment]
This is now fixed with PR #14133.
